R/fit_model.R

```diff
@@ -0,0 +1,174 @@
+#'   EpiNow2. It has required keys: `cores`, `chains`, `iter_warmup`,
+#'   `iter_sampling`, `max_treedepth`, and `adapt_delta`.
+#'
+#' @return A fitted model object of class `epinow` or, if model fitting fails,
+#'   an NA is returned with a warning
+#' @export
+fit_model <- function(
+    data,
+    parameters,
+    seed,
+    horizon,
+    priors,
+    sampler_opts) {
+  # Priors ------------------------------------------------------------------
+  rt <- EpiNow2::rt_opts(
+    list(
+      mean = priors[["rt"]][["mean"]],
+      sd = priors[["rt"]][["sd"]]
+    )
+  )
+  gp <- EpiNow2::gp_opts(
+    alpha_sd = priors[["gp"]][["alpha_sd"]]
+  )
+
+  # Distributions -----------------------------------------------------------
+  generation_time <- format_generation_interval(
+    parameters[["generation_interval"]]
+  )
+  delays <- format_delay_interval(
+    parameters[["delay_interval"]]
+  )
+  truncation <- format_right_truncation(
+    parameters[["right_truncation"]],
+    data
+  )
+
+  # Stan sampler ------------------------------------------------------------
+  stan <- EpiNow2::stan_opts(
+    cores = sampler_opts[["cores"]],
+    chains = sampler_opts[["chains"]],
+    # NOTE: seed gets used twice -- as the seed passed here to the Stan sampler
+    # and below as the R PRNG seed for EpiNow2 initialization
+    seed = seed,
+    warmup = sampler_opts[["iter_warmup"]],
+    samples = sampler_opts[["iter_samples"]],
+    control = list(
+      adapt_delta = sampler_opts[["adapt_delta"]],
+      max_treedepth = sampler_opts[["max_treedepth"]]
+    )
+  )
+
+  df <- data.frame(
+    confirm = data[["confirm"]],
+    date = as.Date(data[["reference_date"]])
+  )
+  rlang::try_fetch(
+    withr::with_seed(seed, {
+      EpiNow2::epinow(
+        df,
+        generation_time = generation_time,
+        delays = delays,
+        truncation = truncation,
+        horizon = horizon,
+        rt = rt,
+        gp = gp,
+        stan = stan,
+        verbose = interactive()
+      )
+    }),
+    # Downgrade model erroring out to a warning so we can catch and return
+    error = function(cnd) {
+      cli::cli_warn(
+        "Model fitting failed. Returning NA.",
+        parent = cnd,
+        class = "failing_fit"
+      )
+      NA
+    }
+  )
+}
+
+#' Format PMFs for EpiNow2
+#'
+#' Format PMFs for use by EpiNow2. Delays or right truncation are optional and
+#' can be skipped by passing an NA.
+#'
+#' @param pmf As returned from [CFAEpiNow2Pipeline::read_disease_parameters()].
+#'   A PMF vector or an NA, if not applying the PMF to the model fit.
+#'
+#' @return An EpiNow2::*_opts() formatted object or NA with a message
+#' @name opts_formatter
+NULL
+
+#' @rdname opts_formatter
+#' @export
+format_generation_interval <- function(pmf) {
+  if (
+    rlang::is_na(pmf) || rlang::is_null(pmf)
+  ) {
+    cli::cli_abort("No generation time PMF specified but is required",
+      class = "Missing_GI"
+    )
+  }
+
+  suppressWarnings({
```

Comment:
I also want to flag this choice as a to-discuss. `EpiNow2` throws a "once every 8 hours" warning about the change in the use of the `max` argument. This warning gets thrown every time on CI and causes a failing R CMD check.

I had trouble getting `rlang::try_fetch()` to downgrade the warning to a message, so I gave up and just suppressed the message.

I think this choice is reasonable because we're running on a pinned version of the package. We should be able to catch any issues in a managed upgrade on CI. But if we for some reason think interactive use is needed with multiple versions of EpiNow2 is needed, we should revisit this choice.

Reply:
That's not super clear to me. Maybe @seabbs knows?

Reply:
So only if we remove it if we decide people have had long enough to see the warning and change their code. 

I am not sure of a better way to handle depreciations like this without harming naive users but open to suggestions

Reply:
Yes seems correct that 1. We don't want to get this warning 2. We do want to get other warnings (probably). As to how to achieve that technically I'm unsure

Reply:
I'll make an issue and we can revisit: https://github.com/CDCgov/cfa-epinow2-pipeline/issues/31